### PR TITLE
Clarify deriving code

### PR DESCRIPTION
**Pull request by nnethercote**
A number of clarifications to the deriving code.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by nnethercote** [comment]
Note: the first seven commits are duplicated from #98758 and will be removed once that PR is merged.

The remaining commits are best reviewed one at a time.

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 1]
```diff
@@ -1490,230 +1447,215 @@ impl<'a> TraitDef<'a> {
...
+                        };
+                        let ident = self.mk_pattern_ident(prefix, i);
+                        let path = ident.with_span_pos(sp);
+                        (
+                            sp,
+                            struct_field.ident,
+                            cx.pat(path.span, PatKind::Ident(binding_mode, path, None)),
+                        )
+                    })
+                    .collect();
```
Comment (klensy) [inline compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1481]: Can `collect()` here be skipped? It didn't used in 3rd match variant, plus in other two cases it immediately consumed.

Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1481]: I will avoid the collection into the `Vec` by changing `pieces` to an iterator. Thanks for the suggestion.

**Comment by nnethercote** [comment]
I have rebased.

@bors r=Mark-Simulacrum

**Comment by nnethercote** [comment]
I fixed the `tidy` error.

@bors r=Mark-Simulacrum

**Pull request merged**